File: collector/sources/http_api.go
// Package sources contains the built-in source types. Adding a new source
// type means adding one file to this package: implement core.Source and
// register a constructor in an init function.
package sources

import (
	"context"
	"crypto/tls"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"strings"
	"time"

	"github.com/pcarpe4/bento/collector/core"
)

// httpAPISource polls one or more JSON HTTP endpoints.
type httpAPISource struct {
	urls     []string
	headers  map[string]string
	username string
	password string
	client   *http.Client
}

func init() {
	core.RegisterSource("http_api", newHTTPAPISource)
}

func newHTTPAPISource(cfg core.Fields) (core.Source, error) {
	urls := cfg.StringList("urls")
	if len(urls) == 0 {
		return nil, fmt.Errorf("config field \"urls\" is required")
	}
	timeout, err := cfg.Duration("timeout", defaultHTTPTimeout)
	if err != nil {
		return nil, err
	}
	return &httpAPISource{
		urls:     urls,
		headers:  cfg.StringMap("headers"),
		username: cfg.String("username", ""),
		password: cfg.String("password", ""),
		client:   newHTTPClient(timeout, cfg.Bool("insecure_skip_verify", false)),
	}, nil
}

func (h *httpAPISource) Collect(ctx context.Context) ([]core.Message, error) {
	var batch []core.Message
	for _, u := range h.urls {
		req, err := http.NewRequestWithContext(ctx, http.MethodGet, u, nil)
		if err != nil {
			return nil, err
		}
		req.Header.Set("Accept", "application/json")
		for k, v := range h.headers {
			req.Header.Set(k, v)
		}
		if h.username != "" {
			req.SetBasicAuth(h.username, h.password)
		}

		res, err := h.client.Do(req)
		if err != nil {
			return nil, fmt.Errorf("request for %v failed: %w", u, err)
		}
		body, err := io.ReadAll(res.Body)
		res.Body.Close()
		if err != nil {
			return nil, fmt.Errorf("failed to read response from %v: %w", u, err)
		}
		if res.StatusCode < 200 || res.StatusCode > 299 {
			return nil, fmt.Errorf("request for %v failed: status %v", u, res.StatusCode)
		}

		for _, msg := range jsonToMessages(body) {
			msg.Meta["url"] = u
			batch = append(batch, msg)
		}
	}
	return batch, nil
}

// jsonToMessages converts a response body into messages: a JSON array yields
// one message per object element, a JSON object yields a single message, and
// anything else is wrapped as {"raw": "..."}.
func jsonToMessages(body []byte) []core.Message {
	var asObject map[string]any
	if err := json.Unmarshal(body, &asObject); err == nil {
		return []core.Message{core.NewMessage(asObject)}
	}
	var asList []any
	if err := json.Unmarshal(body, &asList); err == nil {
		msgs := make([]core.Message, 0, len(asList))
		for _, item := range asList {
			if obj, ok := item.(map[string]any); ok {
				msgs = append(msgs, core.NewMessage(obj))
			} else {
				msgs = append(msgs, core.NewMessage(map[string]any{"value": item}))
			}
		}
		return msgs
	}
	return []core.Message{core.NewMessage(map[string]any{"raw": strings.TrimSpace(string(body))})}
}

const defaultHTTPTimeout = 15 * time.Second

func newHTTPClient(timeout time.Duration, insecure bool) *http.Client {
	transport := http.DefaultTransport.(*http.Transport).Clone()
	if insecure {
		transport.TLSClientConfig = &tls.Config{InsecureSkipVerify: true}
	}
	return &http.Client{Timeout: timeout, Transport: transport}
}

File: collector-rs/src/sources/http_api.rs
use anyhow::{bail, Context, Result};
use async_trait::async_trait;
use serde::Deserialize;

use crate::config::parse_duration;
use crate::message::Message;
use crate::registry::Source;

/// Polls one or more JSON HTTP endpoints.
#[derive(Deserialize)]
struct HttpApiConfig {
    urls: Vec<String>,
    #[serde(default)]
    username: String,
    #[serde(default)]
    password: String,
    #[serde(default)]
    headers: std::collections::BTreeMap<String, String>,
    #[serde(default)]
    insecure_skip_verify: bool,
    #[serde(default = "super::default_timeout")]
    timeout: String,
}

struct HttpApiSource {
    cfg: HttpApiConfig,
    client: reqwest::Client,
}

pub fn new(raw: &serde_yaml::Value) -> Result<Box<dyn Source>> {
    let cfg: HttpApiConfig = serde_yaml::from_value(raw.clone()).context("invalid config")?;
    if cfg.urls.is_empty() {
        bail!("config field \"urls\" is required");
    }
    let client = super::http_client(parse_duration(&cfg.timeout)?, cfg.insecure_skip_verify)?;
    Ok(Box::new(HttpApiSource { cfg, client }))
}

#[async_trait]
impl Source for HttpApiSource {
    async fn collect(&self) -> Result<Vec<Message>> {
        let mut batch = Vec::new();
        for url in &self.cfg.urls {
            let mut req = self.client.get(url).header("Accept", "application/json");
            for (k, v) in &self.cfg.headers {
                req = req.header(k, v);
            }
            if !self.cfg.username.is_empty() {
                req = req.basic_auth(&self.cfg.username, Some(&self.cfg.password));
            }

            let res = req
                .send()
                .await
                .with_context(|| format!("request for {url} failed"))?;
            let status = res.status();
            let body = res
                .bytes()
                .await
                .with_context(|| format!("failed to read response from {url}"))?;
            if !status.is_success() {
                bail!("request for {url} failed: status {status}");
            }

            for mut msg in body_to_messages(&body) {
                msg.meta.insert("url".to_string(), url.clone());
                batch.push(msg);
            }
        }
        Ok(batch)
    }
}

/// Converts a response body into messages: a JSON array yields one message
/// per element, a JSON object yields a single message, and anything else is
/// wrapped as {"raw": "..."}.
fn body_to_messages(body: &[u8]) -> Vec<Message> {
    match serde_json::from_slice::<serde_json::Value>(body) {
        Ok(serde_json::Value::Array(items)) => items
            .into_iter()
            .map(|item| match item {
                obj @ serde_json::Value::Object(_) => Message::new(obj),
                other => Message::new(serde_json::json!({ "value": other })),
            })
            .collect(),
        Ok(obj @ serde_json::Value::Object(_)) => vec![Message::new(obj)],
        Ok(other) => vec![Message::new(serde_json::json!({ "value": other }))],
        Err(_) => {
            let raw = String::from_utf8_lossy(body).trim().to_string();
            vec![Message::new(serde_json::json!({ "raw": raw }))]
        }
    }
}
